### PR TITLE
sqlite3 gem and examples

### PR DESCRIPTION
**Pull request by gintama91**
image

image

### CODE REVIEW

File: lib/constants.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+module Constants
+  LIB_DIR = File.join(Dir.home, "Desktop")
```
Comment (noahgibbs) [inline lib/constants.rb:4]: This is going to vary between Mac, Unix and Windows :-(

Comment (noahgibbs) [reply lib/constants.rb:4]: Oh hey, no, you need that available in the top level of Shoes apps. Ooh... Yeah, okay. Never mind, this is probably the way to go :-/

Comment (noahgibbs) [reply lib/constants.rb:4]: Ah, yeah. Looks like Shoes4 doesn't define it. But Shoes3 has a fairly involved definition:

~~~
require 'tmpdir'

lib_dir = nil
homes = []
homes << [ENV['LOCALAPPDATA'], File.join( ENV['LOCALAPPDATA'], 'Shoes')] if ENV['LOCALAPPDATA']
homes << [ENV['APPDATA'], File.join( ENV['APPDATA'], 'Shoes' )] if ENV['APPDATA']
homes << [ENV['HOME'], File.join( ENV['HOME'], '.shoes' )] if ENV['HOME']
homes.each do |home_top, home_dir|
  next unless home_top
  if File.exists? home_top
    lib_dir = home_dir
    break
  end
end
LIB_DIR = lib_dir || File.join(Dir::tmpdir, "shoes")
~~~

Comment (Schwad) [reply lib/constants.rb:4]: @noahgibbs long-term one thing I'm willing to do, but give me feedback before I write an issue, is detect OS at the top level on boot and be able to have OS-dependent handling in the framework.

This space is _fairly_ familiar to me as in my old job we had conditionals for dozens of setups. e.g. pr here https://github.com/inspec/inspec/pull/5164/files#diff-6d9770e6c14d27a0aee3e8d912b78d13277502d8abaee01394e1c712e2439bec

I think for now we'll not aim to support _everything_ but I could see chonks like:

```
case os
when 'darwin'
  do_the_thing
when 'windows'
  @log("The Thing is not supported yet in windows")
  nil
when 'someting-something-linux'
  @log("The Thing is not supported yet in something-something linux")
  nil
else
  raise "Unknown OS!"
end
```

even as documentation so other devs could easily see state

Comment (Schwad) [reply lib/constants.rb:4]: I may move this forward @noahgibbs - we can revert if you disagree - and I'd be interested if there's another approach you'd like on this constant fixed forward?

Comment (Schwad) [reply lib/constants.rb:4]: i.e. I appreciate there might be a world where we'd revisit the crowded constants name

**Review by Schwad — APPROVED** [review]
Thank you for moving these over. I am excited about sqlite3!

File: lib/constants.rb (hunk 1 above)
Comment (noahgibbs) [reply lib/constants.rb:4]: We could move this forward, but $HOME/Desktop isn't going to exist on our CI servers.

Comment (Schwad) [reply lib/constants.rb:4]: What would be the best way to handle that line for our CI do you think? have this be a skip example for CI or... would there be an alternative that would work conditionally for the CI platform? I'm happy to look into this.

Comment (noahgibbs) [reply lib/constants.rb:4]: We could literally copy the Shoes3 code above. That would be my favourite resolution since it keeps API compatibility.

**Pull request merged**